sys/xtimer/xtimer.c

```diff
@@ -246,14 +247,21 @@ int xtimer_mutex_lock_timeout(mutex_t *mutex, uint64_t timeout)
 {
     xtimer_t t;
     mutex_thread_t mt = { mutex, (thread_t *)sched_active_thread, 0 };
+    int locked = mutex_trylock(mutex);
 
-    if (timeout != 0) {
-        t.callback = _mutex_timeout;
-        t.arg = (void *)((mutex_thread_t *)&mt);
-        _xtimer_set64(&t, timeout, timeout >> 32);
+    if (locked || (timeout == 0)) {
+        return (locked - 1) * ETIMEDOUT;
```

Comment:
I know it's ugly... however with an if branch it generates larger code for most boards:

```C
    if (locked) {
        return 0;
    }
    else if (timeout == 0) {
        return -ETIMEDOUT;
    }
```

here is the size diff: http://pastebin.com/VVPjZQfK (+4 +8 bytes larger for most boards but -10 for avr).

Removing the -ETIMEDOUT and returning -1 instead reduces the code size at least by 4 bytes for most boards (many 10 and even 16). Size diff here: http://pastebin.com/dnG7YCJ5)

```C
    if (locked || (timeout == 0)) {
        return (locked - 1);
    }
```

I would prefer this solution. What do you think?